### PR TITLE
Bugfix/fix changelog join table case

### LINKED ISSUE #106: VersionedKV join table was blocked
On proton 1.3.10, run a join query (versioned_kv join table), got error:
```
'VersionedKV stream' INNER ALL JOIN 'Append stream' is not supported
```

### LINKED ISSUE #118: missing `_tp_delta` when select from view over changelog()
**Describe what's wrong**

```sql
create stream if not exists test14_subquery_4(i float, k1 int, k2 string);
create view if not exists test14_view1_4 as (
    select i, k1, k2 from changelog(test14_subquery_4, k2)
);
create view if not exists test14_view2_4 as (
    select * from test14_view1_4
);
select sec_large(i), k2 from test14_view2_4 group by k2;
```
```
SELECT 
  sec_large(i), k2
FROM 
  test14_view2_4
GROUP BY 
  k2

Query id: 76660320-6ce3-4ab1-8906-6f78da32264d


0 rows in set. Elapsed: 0.009 sec. 

Received exception from server (version 1.3.11):
Code: 47. DB::Exception: Received from localhost:8463. DB::Exception: Missing columns: '_tp_delta' while processing query: 'SELECT 
  test14_view1_4.i, test14_view1_4.k2, _tp_delta
FROM 
  (
  SELECT 
    i, k1, k2
  FROM 
    changelog(default.test14_subquery_4, k2)
) AS test14_view1_4', required columns: 'i' '_tp_delta' 'k2', maybe you meant: ['i','k2']. (UNKNOWN_IDENTIFIER)
```
**How to reproduce**

**Error message and/or stacktrace**

**Additional context**

### PR DESCRIPTION
**Pull request by yl-lisen**
PR checklist:
- Did you run ClangFormat ?
- Did you separate headers to a different section in existing community code base ?
- Did you surround `proton: starts/ends` for new code in existing community code base ?

Please write user-readable short description of the changes:
fix #106  and fix #118 , the change log:
- support handle join result semantic for `stream join table`
- fix input tracking changes implementation
- optimize common logic `isStreamingStorage`

### CODE REVIEW

File: src/Interpreters/Context.h [hunk 1]
```diff
@@ -395,7 +395,7 @@ class Context: public std::enable_shared_from_this<Context>
     /// proton: starts. Parameters for time predicates of main table
     TimeParam time_param;
 
-    using DataStreamSemanticCache = std::unordered_map<std::string, Streaming::DataStreamSemanticEx>;
+    using DataStreamSemanticCache = std::unordered_map<std::string, std::pair<Streaming::DataStreamSemanticEx, bool>>;
```
Comment (sunset3000) [inline src/Interpreters/Context.h]: Why do not put this bool variable into 'DataStreamSemanticEx'?

File: src/Storages/Streaming/storageUtil.cpp [hunk 2]
```diff
@@ -26,4 +29,23 @@ String getStorageName(const ASTCreateQuery & create)
     else
         return "Stream";
 }
+
+bool isStreamingStorage(const StoragePtr & storage, ContextPtr context)
```
Comment (sunset3000) [inline src/Storages/Streaming/storageUtil.cpp:33]: Why do not modify the logic of ProxyStream::isStreaming()? Why an extra function is added here?

File: src/Interpreters/DatabaseAndTableWithAlias.h [hunk 3]
```diff
@@ -59,11 +59,14 @@ struct TableWithColumnNamesAndTypes
...
     /// Physical storage only has output data stream semantic and doesn't have input data stream semantic
     Streaming::DataStreamSemanticEx output_data_stream_semantic;
+    bool is_streaming_output = false;
 
     void setOutputDataStreamSemantic(Streaming::DataStreamSemanticEx data_stream_semantic_)
     {
         output_data_stream_semantic = data_stream_semantic_;
     }
+
+    void setStreamingOutput(bool is_streaming) { is_streaming_output = is_streaming; }
```
Comment (sunset3000) [inline src/Interpreters/DatabaseAndTableWithAlias.h]: if 'is_streaming_output' has been added into 'DataStreamSemanticEx', we do not need this function at all

File: src/Interpreters/Streaming/CalculateDataStreamSemantic.h [hunk 4]
```diff
@@ -13,10 +13,19 @@ namespace Streaming
...
 /// 1) Storage (VersionedKV/ChangelogKV/Changelog)
 /// 2) Or flat transformation / filtering data stream over 1) Storages
-bool canTrackChangesFromStorage(DataStreamSemanticEx input_data_stream_semantic);
+/// 3) Or changelog data stream
+bool canTrackChangesFromInput(DataStreamSemanticEx input_data_stream_semantic);
 
+/// Used for <stream> join <stream>
 bool isJoinResultChangelog(DataStreamSemanticEx left_data_stream_semantic, DataStreamSemanticEx right_data_stream_semantic);
 
+/// Used for <stream/table> join <stream/table>
```
Comment (sunset3000) [inline src/Interpreters/Streaming/CalculateDataStreamSemantic.h]: It is wierd to add this function. Only 'bool isJoinResultChangelog(DataStreamSemanticEx left_data_stream_semantic, DataStreamSemanticEx right_data_stream_semantic)' is enough. It is better to put the logic of this function into the origin 'isJoinResultChangelog' function.

File: src/Interpreters/Streaming/CalculateDataStreamSemantic.h [hunk 5]
```diff
@@ -33,6 +42,7 @@ DataStreamSemanticPair calculateDataStreamSemantic(
     DataStreamSemanticEx left_input_data_stream_semantic,
     std::optional<DataStreamSemanticEx> right_input_data_stream_semantic,
     std::optional<std::pair<JoinKind, JoinStrictness>> kind_and_strictness,
+    std::optional<bool> right_input_is_streaming,
```
Comment (sunset3000) [inline src/Interpreters/Streaming/CalculateDataStreamSemantic.h]: No need to add this param. It is better to put this variable into 'right_input_data_stream_semantic'

File: src/Interpreters/Streaming/HashJoin.cpp [hunk 6]
```diff
@@ -817,18 +817,6 @@ const HashJoin::SupportMatrix HashJoin::support_matrix = {
     {{StorageSemantic::VersionedKV, JoinKind::Inner, JoinStrictness::All, StorageSemantic::ChangelogKV}, true},
     {{StorageSemantic::VersionedKV, JoinKind::Inner, JoinStrictness::All, StorageSemantic::VersionedKV}, true},
     {{StorageSemantic::VersionedKV, JoinKind::Inner, JoinStrictness::All, StorageSemantic::Changelog}, true},
-
-    {{StorageSemantic::VersionedKV, JoinKind::Left, JoinStrictness::Asof, StorageSemantic::Append}, true},
-    {{StorageSemantic::VersionedKV, JoinKind::Left, JoinStrictness::Asof, StorageSemantic::VersionedKV}, true},
-
-    {{StorageSemantic::VersionedKV, JoinKind::Left, JoinStrictness::Any, StorageSemantic::Append}, true},
-    {{StorageSemantic::VersionedKV, JoinKind::Left, JoinStrictness::Any, StorageSemantic::VersionedKV}, true},
```
Comment (sunset3000) [inline src/Interpreters/Streaming/HashJoin.cpp:825]: Why do not support versioned_kv left join versioned_kv?

File: src/Interpreters/Streaming/HashJoin.cpp [hunk 7]
```diff
@@ -817,18 +817,6 @@ const HashJoin::SupportMatrix HashJoin::support_matrix = {
...
-
-    {{StorageSemantic::VersionedKV, JoinKind::Inner, JoinStrictness::Asof, StorageSemantic::Append}, true},
-    {{StorageSemantic::VersionedKV, JoinKind::Inner, JoinStrictness::Asof, StorageSemantic::VersionedKV}, true},
```
Comment (sunset3000) [inline src/Interpreters/Streaming/HashJoin.cpp:828]: Why do not support versioned_kv inner asof join versioned_kv?

File: src/Interpreters/Streaming/HashJoin.cpp [hunk 8]
```diff
@@ -817,18 +817,6 @@ const HashJoin::SupportMatrix HashJoin::support_matrix = {
...
-
-    {{StorageSemantic::VersionedKV, JoinKind::Inner, JoinStrictness::Any, StorageSemantic::Append}, true},
-    {{StorageSemantic::VersionedKV, JoinKind::Inner, JoinStrictness::Any, StorageSemantic::VersionedKV}, true},
```
Comment (sunset3000) [inline src/Interpreters/Streaming/HashJoin.cpp:831]: Why do not support versioned_kv inner any join versioned_kv?

File: src/Interpreters/Streaming/HashJoin.cpp (hunk 6 above)
Comment (yl-lisen) [reply src/Interpreters/Streaming/HashJoin.cpp:825]: Some cases to support

File: src/Interpreters/Streaming/GetSampleBlockContext.h [hunk 9]
```diff
@@ -0,0 +1,16 @@
+#pragma once
+
+#include <Core/Streaming/DataStreamSemantic.h>
+
+namespace DB
+{
+namespace Streaming
+{
+struct GetSampleBlockContext
```
Comment (sunset3000) [inline src/Interpreters/Streaming/GetSampleBlockContext.h]: No need to add this struct and related function. If 'is_streaming_output' put into DataStreamSemanticEx

File: src/Storages/Streaming/storageUtil.cpp (hunk 2 above)
Comment (yl-lisen) [inline src/Storages/Streaming/storageUtil.cpp:33]: There are different storages (ProxyStream is only one of them), such as `Stream', 'Random stream', 'MaterializedView' .etc , in addition, it depends on setting `query_mode` in Context.
So we need a common function to check

File: src/Interpreters/Context.h (hunk 1 above)
Comment (yl-lisen) [reply src/Interpreters/Context.h]: Yes, it's a good way

File: src/Interpreters/Streaming/HashJoin.cpp (hunk 7 above)
Comment (yl-lisen) [reply src/Interpreters/Streaming/HashJoin.cpp:828]: Some cases to support

File: src/Interpreters/Streaming/HashJoin.cpp (hunk 8 above)
Comment (yl-lisen) [reply src/Interpreters/Streaming/HashJoin.cpp:831]: Some cases to support

File: src/Interpreters/InterpreterSelectWithUnionQuery.h [hunk 10]
```diff
@@ -47,7 +47,7 @@ class InterpreterSelectWithUnionQuery : public IInterpreterUnionOrSelectQuery
         const ASTPtr & query_ptr_,
         ContextPtr context_,
         bool is_subquery,
-        Streaming::DataStreamSemanticEx * output_data_stream_semantic = nullptr);
+        Streaming::GetSampleBlockContext * get_sample_block_ctx = nullptr);
```
Comment (sunset3000) [inline src/Interpreters/InterpreterSelectWithUnionQuery.h]: No need to change the signature of this function, is 'is_streaming_output' has been put into DataStreamSemanticEx

File: src/Interpreters/Streaming/CalculateDataStreamSemantic.cpp [hunk 11]
```diff
@@ -13,53 +13,77 @@ DataStreamSemanticEx calculateDataStreamSemanticForJoin(
     DataStreamSemanticEx left_input_data_stream_semantic,
     DataStreamSemanticEx right_input_data_stream_semantic,
     std::pair<JoinKind, JoinStrictness> kind_and_strictness,
+    bool right_input_is_streaming,
     SelectQueryInfo & query_info)
 {
-    HashJoin::validate(
```
Comment (sunset3000) [inline src/Interpreters/Streaming/CalculateDataStreamSemantic.cpp:18]: Why remove the validation here?

Comment (yl-lisen) [reply src/Interpreters/Streaming/CalculateDataStreamSemantic.cpp:18]: The func is used for validate ` stream join stream `, so skipped here. and it's already checked in `Streaming::HashJoin::init()`

**Review by sunset3000 — APPROVED** [review]

File: src/Interpreters/Streaming/CalculateDataStreamSemantic.cpp [hunk 12]
```diff
@@ -68,14 +99,22 @@ DataStreamSemanticEx calculateDataStreamSemanticForJoin(
 }
 }
 
-bool canTrackChangesFromStorage(DataStreamSemanticEx input_data_stream_semantic)
+bool canTrackChangesFromInput(DataStreamSemanticEx input_data_stream_semantic)
 {
-    return isVersionedKeyedStorage(input_data_stream_semantic) || isChangelogKeyedStorage(input_data_stream_semantic)
-        || isChangelogKeyedStorage(input_data_stream_semantic);
+    return isChangelogDataStream(input_data_stream_semantic) || isVersionedKeyedStorage(input_data_stream_semantic)
+        || isChangelogKeyedStorage(input_data_stream_semantic) || isChangelogKeyedStorage(input_data_stream_semantic);
```
Comment (yokofly) [inline src/Interpreters/Streaming/CalculateDataStreamSemantic.cpp:105]: why check 2 times `isChangelogKeyedStorage(input_data_stream_semantic)`

**Pull request merged**